practice/rain.py

```diff
@@ -4,139 +4,99 @@
-    >>> make_lookup_table([['A', 'B']])
-    {'A': 'B'}
+def get_date_to_amt(table):
+    """Turns data into a dictionary that pairs each date with the rainfall amount on that day.
 
-    :param formatted_table:
-    :return:
+    >>> get_date_to_amt([['A', '3']])
+    {'A': 3}
     """
-    lookup_table = {key: value for (key, value) in formatted_table}
-    return lookup_table
+    date_to_amt = {key: value for (key, value) in table}
+    return filter_table(date_to_amt)
+
+def find_date_with_most_rain(date_to_amt):
+    """Compares all dictionary items and returns the date with the highest value.
 
+    >>> find_date_with_most_rain({'A': 1, 'B': 2})
+    'B'
+    """
+    return max(date_to_amt.items(), key=operator.itemgetter(1))[0]
 
+def filter_table(date_to_amt):
+    """Removes dates which have no data from data set.
 
-def find_date_with_most_rain(filtered_table):
+    >>> filter_table({'15-DEC-2010': '-', '24-JUL-2009': '0'})
+    {'24-JUL-2009': 0}
     """
-    >>> find_date_with_most_rain({'A': '1', 'B': '2'})
-    B
+    return {date: int(amt) for date, amt in date_to_amt.items() if not '-' == amt}
+
+def sort_table_by_year(date_to_amt):
+    """Groups all values into new items based on the year of the data.
 
-    :param lookup_table:
-    :return:
+    >>> [(key, sorted(list_val)) for key, list_val in sorted(sort_table_by_year({'15-NOV-2009': 6, '08-JAN-2009': 0}).items())]
+    [('2009', [0, 6])]
     """
-    date_with_most_rain = max(filtered_table.items(), key=operator.itemgetter(1))[0]
-    print(date_with_most_rain)
-    return date_with_most_rain
-
-def filter_table(lookup_table): # Check to make sure this works
-    days_without_value = find_days_without_value(lookup_table)
-    corrected_days = correct_days_without_value(days_without_value)
-    lookup_table.update(corrected_days)
-    return lookup_table
-
-# def filter_table(lookup_table):
-#     """
-#     >>> filter_table({'15-DEC-2010': '-', '24-JUL-2009': '0'})
-#     {'A': '1'}
-#
-#     :param lookup_table:
-#     :return:
-#     """
-#     days_without_value = find_days_without_value(lookup_table)
-#     filtered_table = {lookup_table.pop(item) for item in list(lookup_table) if item in list(days_without_value.keys())}
-#     return filtered_table
-
-
-def find_days_without_value(lookup_table):
+    year_to_amts= {}
+    for date, amt in date_to_amt.items():
+        year = date[-4:]
+        if year not in year_to_amts:
+            year_to_amts[year] = []
+        year_to_amts[year] += [amt]
+    return year_to_amts
+
+def find_year_with_most_rain(year_to_amts):
+    """Compares all items and returns the year with the highest value.
+
+    >>> find_year_with_most_rain({'2011': [0, 1, 2, 3], '2012': [2, 3, 4, 5]})
+    '2012'
     """
-    >>> find_days_without_value({'15-DEC-2010': '-', '24-JUL-2009': '0'})
-    {'B': '-'}
+    total_rain_by_year = sum_yearly_amounts(year_to_amts)
+    return max(total_rain_by_year.items(), key=operator.itemgetter(1))[0]
```

Comment:
If you only care about the max _value_, do something like

```py
max(total_rain_by_year.values(), key=total_rain_by_year.get)
```

Then you don't need to unpack and pack.